extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentPusher.java

```diff
@@ -230,4 +231,22 @@ public String makeIndexPathName(DataSegment dataSegment, String indexName)
         indexName
     );
   }
+
+
+  // We lazily initialiize fullQualifiedStorageDirectory to avoid potential issues with Hadoop namenode HA.
+  // Please see https://github.com/druid-io/druid/pull/5684
+  private void initFullyQualifiedStorageDirectory()
+  {
+    try {
+      if (fullyQualifiedStorageDirectory == null) {
```

Comment:
If it was a conscious decision to not synchronize this, say why in a comment. Or if it wasn't a conscious decision, consider synchronizing it.

Reply:
Maybe https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Suppliers.html#memoize(com.google.common.base.Supplier)? Might be cleaner than calling init in places too

Reply:
Went with the memoize approach